notebooks/Batchelor.ipynb

```diff
@@ -0,0 +1,348 @@
+   "id": "2d81ad31-0597-42d4-ad0e-ac770045eedb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def batchelor(ne, p=1):\n",
+    "    # define the domain (unit square) of length and width ne, running in parallel\n",
+    "    domain = mesh.create_unit_square(MPI.COMM_WORLD, ne, ne)\n",
+    "\n",
+    "    # define velocity and pressure element\n",
+    "    v_e = basix.ufl.element(\"Lagrange\", domain.basix_cell(), p+1, shape=(domain.geometry.dim,))\n",
+    "    p_e = basix.ufl.element(\"Lagrange\", domain.basix_cell(), p)\n",
+    "\n",
+    "    # define mixed element\n",
+    "    vPe = basix.ufl.mixed_element([v_e, p_e])\n",
+    "\n",
+    "    # define function space\n",
+    "    V = functionspace(domain, vPe)\n",
+    "\n",
+    "    # define Velocity and Pressure function space\n",
+    "    V_v, _ = V.sub(0).collapse()\n",
+    "    V_p, _ = V.sub(1).collapse()\n",
+    "\n",
+    "    # define test functions\n",
+    "    v_t, p_t = ufl.TestFunctions(V)\n",
+    "\n",
+    "    # define trial functions\n",
+    "    trial_v, trial_p = ufl.TrialFunctions(V)\n",
+    "\n",
+    "    # define boundary value for Dirichlet boundary condition (59)\n",
+    "    def boundary_D_1(x):\n",
+    "        return np.isclose(x[0], 0)\n",
+    "\n",
+    "    dofs_D_1 = locate_dofs_geometrical((V.sub(0), V_v), boundary_D_1)\n",
+    "    u_bc1 = Function(V_v)\n",
+    "    u_bc1.x.array[:] = 0\n",
+    "    bc1 = dirichletbc(u_bc1, dofs_D_1, V.sub(0))\n",
+    "\n",
+    "    # define boundary value for Dirichlet boundary condition (60)\n",
+    "    def boundary_D_2(x):\n",
+    "        return np.isclose(x[1], 0)\n",
+    "\n",
+    "    dofs_D_2 = locate_dofs_geometrical((V.sub(0), V_v), boundary_D_2)\n",
+    "    u_bc2 = Function(V_v)\n",
+    "    u_bc2.interpolate(u_0)\n",
+    "    bc2 = dirichletbc(u_bc2, dofs_D_2, V.sub(0))\n",
+    "\n",
+    "    # define boundary value for Dirichlet boundary condition (61)\n",
+    "    def boundary_D_3(x):\n",
+    "        return np.logical_or(np.isclose(x[0], 1), np.isclose(x[1], 1))\n",
+    "\n",
+    "    dofs_D_3 = locate_dofs_geometrical((V.sub(0), V_v), boundary_D_3)\n",
+    "    u_bc3 = Function(V_v)\n",
+    "    u_bc3.interpolate(u_exact_np)\n",
+    "    bc3 = dirichletbc(u_bc3, dofs_D_3, V.sub(0))\n",
+    "\n",
+    "    def boundary_D_4(x):\n",
+    "        return np.logical_and(np.isclose(x[0], 0), np.isclose(x[1], 0))\n",
+    "\n",
+    "    dofs_D_4 = locate_dofs_geometrical((V.sub(1), V_p), boundary_D_4)\n",
+    "    u_bc4 = Function(V_p)\n",
+    "    u_bc4.x.array[:] = 0\n",
+    "    bc4 = dirichletbc(u_bc4, dofs_D_4, V.sub(1))\n",
+    "\n",
+    "    # assemble weak form\n",
+    "    K = ufl.inner(ufl.sym(ufl.grad(v_t)), ufl.sym(ufl.grad(trial_v))) * ufl.dx\n",
+    "    G = -ufl.div(v_t)*trial_p*ufl.dx\n",
+    "    D = -p_t*ufl.div(trial_v)*ufl.dx\n",
+    "    S = K + G + D\n",
+    "\n",
+    "    zeroes = Constant(domain, default_scalar_type(0.0))\n",
```

Comment:
Could this be called `zero` or `zero_c` for consistency and because it isn't zeros plural :-)